File: 176_MaxOfMinWindow.py
"""
Problem: Max of Min of Every Window Size
Description: For every window size, find maximum of minimums in all windows of that size.

Approach: Use stack to find next and previous smaller elements. Calculate window size for each element as minimum.
"""

from typing import List

def maxOfMin(arr: List[int]) -> List[int]:
    n = len(arr)
    left = [0] * n
    right = [0] * n
    stack = []
    
    # Find previous smaller
    for i in range(n):
        while stack and arr[stack[-1]] >= arr[i]:
            stack.pop()
        left[i] = -1 if not stack else stack[-1]
        stack.append(i)
    
    stack.clear()
    
    # Find next smaller
    for i in range(n - 1, -1, -1):
        while stack and arr[stack[-1]] >= arr[i]:
            stack.pop()
        right[i] = n if not stack else stack[-1]
        stack.append(i)
    
    result = [0] * (n + 1)
    
    # Fill result
    for i in range(n):
        length = right[i] - left[i] - 1
        result[length] = max(result[length], arr[i])
    
    # Fill remaining positions
    for i in range(n - 1, 0, -1):
        result[i] = max(result[i], result[i + 1])
    
    return result[1:]

def main():
    arr = [10, 20, 30, 50, 10, 70, 30]
    result = maxOfMin(arr)
    
    print(f"Max of min for each window size: {result}")
    # Output: [70, 30, 20, 10, 10, 10, 10]

if __name__ == "__main__":
    main()

"""
Time Complexity: O(n)
Space Complexity: O(n)
"""

File: 176_MaxOfMinWindow.cpp
/*
Problem: Max of Min of Every Window Size
Description: For every window size, find maximum of minimums in all windows of that size.

Approach: Use stack to find next and previous smaller elements. Calculate window size for each element as minimum.
*/

#include <bits/stdc++.h>
using namespace std;

vector<int> maxOfMin(vector<int>& arr) {
    int n = arr.size();
    vector<int> left(n), right(n);
    stack<int> st;
    
    // Find previous smaller
    for (int i = 0; i < n; i++) {
        while (!st.empty() && arr[st.top()] >= arr[i]) {
            st.pop();
        }
        left[i] = st.empty() ? -1 : st.top();
        st.push(i);
    }
    
    while (!st.empty()) st.pop();
    
    // Find next smaller
    for (int i = n - 1; i >= 0; i--) {
        while (!st.empty() && arr[st.top()] >= arr[i]) {
            st.pop();
        }
        right[i] = st.empty() ? n : st.top();
        st.push(i);
    }
    
    vector<int> result(n + 1, 0);
    
    // Fill result
    for (int i = 0; i < n; i++) {
        int len = right[i] - left[i] - 1;
        result[len] = max(result[len], arr[i]);
    }
    
    // Fill remaining positions
    for (int i = n - 1; i >= 1; i--) {
        result[i] = max(result[i], result[i + 1]);
    }
    
    vector<int> ans(result.begin() + 1, result.end());
    return ans;
}

int main() {
    vector<int> arr = {10, 20, 30, 50, 10, 70, 30};
    vector<int> result = maxOfMin(arr);
    
    cout << "Max of min for each window size: ";
    for (int x : result) {
        cout << x << " ";
    }
    cout << endl;
    // Output: 70 30 20 10 10 10 10
    
    return 0;
}

/*
Time Complexity: O(n)
Space Complexity: O(n)
*/
